IntegrationTest/ExerciseEndpointsTest.cs

```diff
@@ -0,0 +1,576 @@
+﻿using API;
+using Core.Exercises.Contracts;
+using Core.Exercises.Models;
+using Core.Shared;
+using Core.Solutions.Contracts;
+using Core.Solutions.Models;
+using Core.Solutions.Services;
+using FluentResults;
+using IntegrationTest.Setup;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using NSubstitute;
+using System.Net;
+using System.Net.Http.Json;
+using System.Text;
+
+namespace IntegrationTest;
+
+[Collection(CollectionDefinitions.NonParallelCollectionName)]
+public class ExerciseEndpointsTest : IClassFixture<TestWebApplicationFactory<Program>>
+{
+    private readonly HttpClient _client;
+    private readonly TestWebApplicationFactory<Program> _factory;
+    private readonly ILogger<HaskellService> haskellLoggerSub = Substitute.For<ILogger<HaskellService>>();
+
+    public ExerciseEndpointsTest(TestWebApplicationFactory<Program> factory)
+    {
+        _factory = factory;
+        _client = factory.CreateClient();
+    }
+
+    [Fact]
+    public async Task CreateExercise_ShouldReturn_201()
+    {
+        using var scope = _factory.Services.CreateScope();
+        var haskellServiceSub = scope.ServiceProvider.GetService<IHaskellService>();
+        var exerciseRepoSub = scope.ServiceProvider.GetService<IExerciseRepository>();
+
+        var solutionRunnerResponse = new SolutionRunnerResponse { Action = ResponseCode.Pass };
+        haskellServiceSub!.SubmitSubmission(Arg.Any<SubmissionDto>()).Returns(solutionRunnerResponse);
+        exerciseRepoSub!.InsertExerciseAsync(Arg.Any<ExerciseDto>(), Arg.Any<int>()).Returns(Result.Ok());
+
+        var userId = 1;
+        var roles = new List<Roles> { Roles.Instructor };
+        _client.AddRoleAuth(userId, roles);
+        var requestBody = CreateExerciseRequestBody();
```

Comment:
Good idea using a helper for the request body if it is used multiple times, however if you want you can avoid manually serializing the object and just use the object by changing _client.PostAsync -> _client.PostAsJsonAsync